src/components/organisms/Badges/Badges.tsx

```diff
@@ -0,0 +1,165 @@
+      .reduce((acc, v) => acc + v.timeSpent, 0);
+    return Math.floor(visitSeconds / (60 * 60));
+  }, [visits, currentVenue, relatedVenueIds]);
+
+  // Only show visits to related venues
+  const relevantVisits = visits?.filter(
+    (visit) =>
+      currentVenue.id === visit.id || relatedVenueIds.includes(visit.id)
+  );
+
+  const badges = useMemo(() => {
+    if (!relevantVisits || !venues) return [];
+    return relevantVisits
+      .map((visit) => {
+        const { venue, room } = findVenueAndRoomByName(visit.id, venues);
+        if (
+          !venue ||
+          (currentVenue.id !== venue.id && !relatedVenueIds.includes(venue.id))
+        )
+          return undefined;
+
+        if (room) {
+          return {
+            venue,
+            room,
+            image: room.image_url,
+            label: room.title,
+          };
+        }
+
+        return {
+          venue,
+          image: venue.mapIconImageUrl,
+          label: venue.name,
+        };
+      })
+      .filter((b) => b !== undefined);
+  }, [relevantVisits, venues, currentVenue, relatedVenueIds]);
+
+  if (!relevantVisits) {
+    return <>Visit venues to collect badges!</>;
+  }
+
+  return (
+    <>
+      <div className="visits">
+        <div className="visit-item">
+          <span className="visit-item__value">
+            {visitHours > 1 ? `${visitHours}` : "< 1"} hrs
+          </span>
+          <span className="visit-item__label">Time spent in Sparkle</span>
+        </div>
+        <div className="separator"></div>
+        <div className="visit-item">
+          <span className="visit-item__value">
+            {relevantVisits?.length ?? 0}
+          </span>
+          <span className="visit-item__label">Venues visited</span>
+        </div>
+      </div>
+      <div className="badges-container">
+        <div className="badges-title">{badges.length} Badges</div>
+        <ul className="badge-list">
+          {badges.map(
+            (b) =>
+              b && (
+                <li className="badge-list-item" key={b.label}>
+                  <Link to={getLocationLink(b.venue, b.room)}>
+                    <img
+                      className="badge-list-item-image"
+                      src={b.image}
+                      alt={`${b.label} badge`}
+                    />
+                  </Link>
+                </li>
+              )
+            // label missing on hover - similar to playa
+          )}
+        </ul>
+      </div>
+    </>
+  );
+};
+
+const findVenueAndRoomByName = (
+  nameOrRoomTitle: string,
+  venues: Array<WithId<AnyVenue>>
+) => {
+  const venue = venues.find(
+    (v) =>
+      v.name === nameOrRoomTitle ||
+      (isCampVenue(v) && v.rooms.find((r) => r.title === nameOrRoomTitle))
+  );
+
+  if (!venue) return {};
+
+  if (venue.name === nameOrRoomTitle) return { venue };
+
+  return {
+    venue,
+    room:
+      isCampVenue(venue) &&
+      venue.rooms.find((r) => r.title === nameOrRoomTitle),
+  };
```

Comment:
This is the same logic as used in the `const venue` above isn't it? Can we de-dupe and use that in both places?

Reply:
I had a shot at this, but there's some bad typing in the `rooms` property on the different templates. That will all go away with #833 so going to leave it for now :(

Reply:
Makes sense.